AWS deployment - port to terraform 0.12

flavio opened the pull request:
## Why is this PR needed?

We are switching to terraform 0.12 to deploy all our infrastructure. With AWS is even more important because the AWS provider we've packaged is built to support only terraform 0.12.

## What does this PR do?

It's a port of the existing terraform state files to the new 0.12 syntax.

## Anything else a reviewer needs to know?

The PR consists of some commits, each one of them is isolated from the others. I suggest to take a look at the commits on an individual basis to better understand what they are doing.

On top of the mere 0.12 migration I also did the following improvements:

  * Take advantage of some features of 0.12 to clean up the code
  * Remove a variable that wasn't actually used
  * Add more tags to some of the network resources we're creating. I was debugging some issue with @bergmannf and the lack of these tags was making our life miserable.

## Info for QA

Nothing special to be done. The AWS feature is going to be released as tech preview.

### Status **BEFORE** applying the patch

The AWS infra was deployable only with with terraform <= 0.11

### Status **AFTER** applying the patch

The AWS infra is deployable only with terraform >= 0.12. Note well: there's no need to be retro-compatible with this platform because we haven't released that yet. Hence we can make terraform 0.12 the baseline.

## Docs

Nothing ot be done. There's already a PR about AWS doc.

# Merge restrictions

(Please do not edit this)

We are in *v4-maintenance phase*, so we will restrict what can be merged to prevent unexpected surprises:

    What can be merged (merge criteria):
        2 approvals:
            1 developer: code is fine
            1 QA: QA is fine
        there is a PR for updating documentation (or a statement that this is not needed)

<!-- Remember, if this is a work in progress please pre-append [WIP] to the title until you are ready! 
    If you can, please apply all applicable labels to help reviews out! -->

flavio commented:
**Note well:** the CI checks are failing because the terraform linter is NOT using the 0.12 version, hence it chokes over the new 0.12 syntax.

ereslibre commented:
When smoke testing the AWS provider it wasn't working as expected on 0.12. This was never merged.

@evrardjp can you please take this into account on the merging queue? Thank you!

evrardjp commented:
I will be waiting for the first tests of QA to then move this forward with the other patch.

flavio commented:
I've tested the PR with @evrardjp's terraform-aws package and everything works as expected :clap: 

I've rebased the PR against master and I've also added a minor change to the example terraform tfvars file. I think we are good to go

jordimassaguerpla commented:
> I've tested the PR with @evrardjp's terraform-aws package and everything works as expected clap
> 
> I've rebased the PR against master and I've also added a minor change to the example terraform tfvars file. I think we are good to go

Thanks Flavio. We are going to update the CI to tf0.12 once we have the tf0.12 package in SLE15SP1. All this should happen in this sprint. Please hold this until the tests pass (=tf0.12 is updated) .

davidcassany commented:
> Thanks Flavio. We are going to update the CI to tf0.12 once we have the tf0.12 package in SLE15SP1. All this should happen in this sprint. Please hold this until the tests pass (=tf0.12 is updated) .

I'd wait until the `terraform-provider-aws` package is ready, not only tf012 package. Note that the submission to SLE of aws provider has not yet started, it will happen later than tf012.

jordimassaguerpla commented:
Hi!
New code has been merged in master regarding terraform 0.12. Please rebase your branch to take this changes in.
Happy hacking

flavio commented:
I've rebased the code against master

evrardjp commented on ci/infra/aws/network.tf:
```diff
@@ -1,8 +1,14 @@
+  tags = merge(
+    local.tags,
+    {
+      "Name"  = "${var.stack_name}-vpc"
```
:+1:

evrardjp commented on ci/infra/aws/network.tf:
```diff
@@ -13,94 +19,137 @@ data "aws_availability_zones" "az" {
-  depends_on = ["aws_internet_gateway.platform"]
+  depends_on = [aws_internet_gateway.platform]
+
+  tags = merge(
```
Nice addition.

flavio commented:
Tested on AWS, it works.

flavio commented:
> I am not familiar with tf, nor its aws plugin (and therefore what it exposes as local. variable). I am wondering why we are sometimes merging `local.tags` with our extra data, and sometimes `local.basic_tags`. Shouldn't we simply pass our data in tags like for example, for security_groups:
> `tags = { "Name" = "${var.stack_name}-master" "Class" = "SecurityGroup" }`
> ?

There's a distinction between `basic_tags` and `local` tags because the kubernetes CPI is really picky about the tags associated with the AWS resources. They are used to identify what is managed by the CPI and what not.

I would prefer to not touch that right now, we can refactor the code later if we want to. But changing that right now would put the CPI enablement at risk

flavio commented:
> I am not familiar with tf, nor its aws plugin (and therefore what it exposes as local. variable). I am wondering why we are sometimes merging `local.tags` with our extra data, and sometimes `local.basic_tags`. Shouldn't we simply pass our data in tags like for example, for security_groups:
> `tags = { "Name" = "${var.stack_name}-master" "Class" = "SecurityGroup" }`
> ?

There's a distinction between `basic_tags` and `local` tags because the kubernetes CPI is really picky about the tags associated with the AWS resources. They are used to identify what is managed by the CPI and what not.

I would prefer to not touch that right now, we can refactor the code later if we want to. But changing that right now would put the CPI enablement at risk

jordimassaguerpla commented:
Tests area failing but the files in the PR are about aws, which I am not aware of any test that tests that, so I would bet the failing tests are not related to aws.

flavio commented:
Yeah, the failures are completely unrelated with that. I'm going to forcefully merge it.

SheilaghM commented:
Why are you planning to force merge?  Is this needed so urgently that you cannot wait for two reviews?  Both Marek and Ty have been looking at it yesterday and today.  We have the 2-review rule for a reason, please set a good example.

evrardjp commented:
@SheilaghM FYI there were github issues yesterday, which delayed some reviews (one landed a few hours after the merge, though it was reviewed before the merge). I didn't know Marek and Ty were looking at it though....

flavio commented:
@evrardjp I've already clarified the situation last night with @SheilaghM, @Klaven and @tdaines42.

For all the ones interested, the PR was already reviewed and approved sometimes ago, I invalidated the approval when I rebased against master.
The failures that happened on the CI were unrelated with the code.

I forcefully merged the PR because of these reasons and because we want to push this AWS integration out since quite some time.

jordimassaguerpla commented:
Sorry @Klaven and @tdaines42 . I am the one who reached out to you for help. I thought @flavio was not available and this was blocking the 4.1.2 release.
